### PR TITLE
Convert FacesJS site to React, add lots of Gallery functionality

### PR DESCRIPTION
**Pull request by tomkennedy22**
This is a fairly sizable diff that impacts a majority of the FacesJS site. The FacesJS core functionality - generate, display, etc, are somewhat changed, and I'll detail changes below, but not huge. 

**As of PR creation date [4/7/2024], this PR is NOT ready to land. Simply creating to spur conversation over features.** 

**TODO:**
1. Make site look better on mobile screens
2. Make sure CLI still works
3. Figure out bundling, building, deploying, prod run, etc

## Overall vision
**FacesJS is great, and having a more interactive editor might bring more users to the site and engage more**

- By adding React, we can give a more elegant user experience
- By allowing the user to see potential features, they can better customize their Face
- Create a closer link between the FacesJS home page & the editor (currently small link at the bottom)

### Core Infra for Changes

- React
- [Vite](https://vitejs.dev/)
- [NextUI](https://nextui.org/) for components
- [Tailwind](https://tailwindcss.com/) for CSS framework
- [Zustand](https://zustand-demo.pmnd.rs/) for state storage
- [Phosphor](https://phosphoricons.com/) for icons
- [Jest](https://jestjs.io/) for testing
- Other stuff for routing, testing,

### CODE REVIEW

File: src/utils.ts [hunk 1]
```diff
@@ -0,0 +1,235 @@
+export const get_from_dict = (obj: object, key: string): any => {
+    let keyParts: string[] = String(key).split('.');
```
Comment (dumbmatter) [inline src/utils.ts:2]: Would be better as:

`const keyParts = key.split(".");`

`key` is already a string, and TypeScript knows that String.split returns string[].

Extra type annotations just make things look more confusing IMHO.

File: src/utils.ts [hunk 2]
```diff
@@ -0,0 +1,235 @@
...
+    let current: any = obj;
+
+    for (let i = 0; i < keyParts.length; i++) {
+        // @ts-ignore
+        const part: string = keyParts[i];
```
Comment (dumbmatter) [inline src/utils.ts:7]: `: string` not needed, `// @ts-ignore` not needed.

File: src/utils.ts [hunk 3]
```diff
@@ -0,0 +1,235 @@
...
+        }
+    }
+
+    return current;
+}
+
+
+export const set_to_dict = (container: { [key: string]: any } | Map<any, any>, key: string, value: any) => {
+    key = `${key}`.trim();
+    const keys: string[] = key.split('.');
```
Comment (dumbmatter) [inline src/utils.ts:32]: Same

File: src/utils.ts [hunk 4]
```diff
@@ -0,0 +1,235 @@
...
+    let current_container = container;
+
+    for (let i = 0; i < keys.length; i++) {
+        // @ts-ignore
+        const current_key: string = keys[i];
```
Comment (dumbmatter) [inline src/utils.ts:37]: Same

File: src/utils.ts [hunk 5]
```diff
@@ -0,0 +1,235 @@
...
+            return null;
+        }
+    }
+
+    return current;
+}
+
+
+export const set_to_dict = (container: { [key: string]: any } | Map<any, any>, key: string, value: any) => {
+    key = `${key}`.trim();
```
Comment (dumbmatter) [inline src/utils.ts:31]: `key = key.trim();`

File: public/Face.tsx [hunk 6]
```diff
@@ -0,0 +1,68 @@
...
+};
+
+const faceToSvgStringMemoized = memoizeWithDeepComparison(faceToSvgString);
+// Set to false to disable memoization
+// Memoization is useful when rendering multiple faces with the same configuration
+//      It will speed up rendering by reusing the same SVG string
+const MEMOIZE_FACE = true;
+
+
+export const Face: React.FC<{ faceConfig: FaceConfig, overrides?: Overrides, maxWidth?: number, width?: number, className?: string }> = ({ faceConfig, overrides, maxWidth, width, className }) => {
```
Comment (dumbmatter) [inline public/Face.tsx:40]: `React.FC` type should basically never be used, you can write this instead as:

```
export const Face = ({ faceConfig, overrides, maxWidth, width, className }: { faceConfig: FaceConfig, overrides?: Overrides, maxWidth?: number, width?: number, className?: string }) => {
```

File: tools/process-svgs.js [hunk 7]
```diff
@@ -1,81 +1,83 @@
...
-if (require.main === module) {
-  processSVGs();
-} else {
-  module.exports = processSVGs;
-}
+// if (require.main === module) {
+//     processSVGs();
+// } else {
+//     module.exports = processSVGs;
+// }
```
Comment (dumbmatter) [inline tools/process-svgs.js:81]: Deleting this means that running `node tools/process-svgs.js` will no longer do anything. idk if there is a way to do this same logic with ESM, but if not, we probably need to separate this into two files (one with just the module export, and one that runs on the CLI).

File: tools/process-svgs.d.ts
Comment (dumbmatter) [inline tools/process-svgs.d.ts:1]: Is this needed? Does any TypeScript file actually import process-svgs?

File: vite.config.ts [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+export default defineConfig({
+	plugins: [
+		react(),
+		{
+			name: 'watch-for-svg-changes',
+			configureServer(server) {
+				let processSVGs;
+				if (typeof window === 'undefined') {
+					processSVGs = require('./tools/process-svgs.js').processSVGs;
+				}
```
Comment (dumbmatter) [inline vite.config.ts:15]: What's the purpose of conditionally importing this?

File: vite.config.ts [hunk 9]
```diff
@@ -0,0 +1,42 @@
...
+					.on("all", (event, filePath) => {
+						console.log(`Detected change in SVGs: ${event} ${filePath}`);
+						if (timeoutId !== null) {
+							clearTimeout(timeoutId);
+						}
+
+						timeoutId = setTimeout(() => {
+							console.log('Processing SVGs due to recent changes');
+							processSVGs();
+						}, delayMs);
```
Comment (dumbmatter) [inline vite.config.ts:33]: Why is this needed? Shouldn't the old code from watch.js work the same here?

File: tsconfig.json [hunk 10]
```diff
@@ -1,12 +1,39 @@
...
+		"noImplicitReturns": true,
+		"noPropertyAccessFromIndexSignature": true,
+		"noUncheckedIndexedAccess": true
+	},
+	"include": [
+		"src",
+		"e2e",
+		"public",
+		"tools",
+	]
```
Comment (dumbmatter) [inline tsconfig.json:38]: Some of these don't exist

File: public/Home.tsx
Comment (dumbmatter) [inline public/Home.tsx:1]: Don't change anything now, just thinking...

Is it actually desirable to have this page in React? Would it be better to have it as plain HTML/CSS/JS, for SEO/performance? And then only have the editor in React.

File: vite.config.ts [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+
+				const delayMs = 1_000;
+				let timeoutId: NodeJS.Timeout | undefined = undefined;
+
+				chokidar
+					.watch([path.join(__dirname, "svgs")], {
+						ignoreInitial: false,
+					})
+					.on("all", (event, filePath) => {
+						console.log(`Detected change in SVGs: ${event} ${filePath}`);
```
Comment (dumbmatter) [inline vite.config.ts:25]: This and the other console.log should only run when a file is actually changed, not on the initial build. Otherwise it fills the terminal and you can't even see the link from Vite to the dev page!

File: public/EditorPage.tsx [hunk 12]
```diff
@@ -0,0 +1,848 @@
...
+                    {(_) => (
+                        <>
+                            <ModalHeader className="flex flex-col gap-1">Paste JSON to Render Face</ModalHeader>
+                            <ModalBody>
+                                <Textarea
+                                    value={textAreaValue}
+                                    isInvalid={!textAreaValid}
+                                    ref={textRef}
+                                    // errorMessage={!textAreaValid ? ("Invalid JSON. Refer to the <a src='https://www.json.org/json-en.html'>JSON spec</a>") : null}
+                                    errorMessage={!textAreaValid ? errorMessage : null}
```
Comment (dumbmatter) [inline public/EditorPage.tsx:796]: For error handling here, get rid of the "Validate" button. Just make validation run when you click "Draw" and keep the window open if it's not valid. I don't think anybody will care about validation errors until they're ready to draw.

**Review by dumbmatter — CHANGES_REQUESTED** [review]
Okay I'm going to request a lot of changes, but the core functionality is awesome!

If at any point you're tired of my change requests and want to stop working on it, let me know and I can just take over. But otherwise, I'm going to assume you want to see it through to the end :)

Can you merge with my master branch to fix the merge conflicts?

High level comments from using the editor a bit:

- Sometimes there are too many options displayed, like for Eye Angle it shows me more options than fit on my monitor and the changes are very slight.

- The UI you get from clicking on a category header is really cool! Maybe that could be the default view, but with everything shown all at once? Then clicking a header could switch to that view and scroll to the correct section.

- The previous point would also help on mobile. Then you could just hide the left menu on mobile, since everything would be on the main screen.

- New version is missing some of the randomization flexibility that the old one had - checking only some things to randomize, and specifying race/gender. Maybe could be added back?

- Download as PNG is really nice. While we're at it, maybe make it support download as SVG too.

- Top bar should probably be fixed position, even when you scroll.

- Top bar buttons are maybe not as prominent as in the old version, I think some people might not notice they are there. Maybe they should be bigger? Colored? Include text labels? Or inline (under the face)  like where some are in the current version? Not sure.

File: tsconfig.json (hunk 10 above)
Comment (tomkennedy22) [reply tsconfig.json:38]: Removing e2e, other three should be included (I think?)

File: vite.config.ts (hunk 9 above)
Comment (tomkennedy22) [reply vite.config.ts:33]: Removing timeoutID - added this because having many SVG changes at once could cause many processSVG() executions at once. This was largely an issue when I kept the code below as False, but I'll flip it to True

`.watch([path.join(__dirname, "svgs")], {
						ignoreInitial: true,
					})`

File: svgs/head/head1.svg
[the remark this answers is not in the document]
Comment (tomkennedy22) [reply svgs/head/head1.svg:1]: Whoops, yeah old change that I forgot to revert

File: package-lock.json
[the remark this answers is not in the document]
Comment (tomkennedy22) [reply package-lock.json:1]: switched over to yarn. I was using a mix of npm/pnpm, but yarn is working via testing now, so all good

File: src/utils.ts [hunk 13]
```diff
@@ -0,0 +1,235 @@
+export const get_from_dict = (obj: object, key: string): any => {
```
[the remark this answers is not in the document]
Comment (tomkennedy22) [reply src/utils.ts:1]: Makes sense, forgot to do so

**Comment by tomkennedy22** [comment]
Changed gallery range generation, so scalable features (eyeAngle, mouthSize, bodySize, etc) generate 11 options, and then sliderScale has 101 options 

```
            if (!rangeConfig.step || !rangeConfig.sliderStep) {
                let range = rangeConfig.max - rangeConfig.min;
                rangeConfig.step = roundTwoDecimals(range / 10);
                rangeConfig.sliderStep = Math.max(roundTwoDecimals(range / 100), 0.01);
            }

            itemConfig.renderOptions.valuesToRender = generateRangeFromStep(rangeConfig.min, rangeConfig.max, rangeConfig.step)
```

For instance, Body Size has range 0.8 ➡️ 1.2. So the available options will be 0.8, 0.84, 0.88, 0.92, 0.96, 1.0, 1.04, 1.08, 1.12, 1.16, 1.20.

File: src/utils.ts (hunk 2 above)
Comment (tomkennedy22) [reply src/utils.ts:7]: Changed from 

```
for (let i = 0; i < keyParts.length; i++) {
        // @ts-ignore
        const part: string = keyParts[i];
```

to

`for (let part of keyParts) {`

File: src/utils.ts (hunk 4 above)
Comment (tomkennedy22) [reply src/utils.ts:37]: Fixed (changed)

File: src/utils.ts (hunk 3 above)
Comment (tomkennedy22) [reply src/utils.ts:32]: Fixed

File: src/utils.ts (hunk 5 above)
Comment (tomkennedy22) [reply src/utils.ts:31]: Combined with line under

File: tools/process-svgs.d.ts
Comment (tomkennedy22) [reply tools/process-svgs.d.ts:1]: Removed

File: public/Face.tsx (hunk 6 above)
Comment (tomkennedy22) [reply public/Face.tsx:40]: Gotcha, fixed

File: src/utils.ts (hunk 2 above)
Comment (dumbmatter) [reply src/utils.ts]: Oh yeah, that's even better. Also personally I prefer using `const` rather than `let` where possible because it's a little clearer about what the variable is for and sometimes it matters for TypeScript too, but I don't care enough to make you do it everywhere :)

File: .husky/.gitignore
Comment (dumbmatter) [inline .husky/.gitignore:1]: Don't change this file. Same with pre-commit (missing the newline at the end of the file)

File: __tests__/utils.get_from_dict.test.ts
Comment (dumbmatter) [inline __tests__/utils.get_from_dict.test.ts:1]: Update filename to match function

File: __tests__/utils.set_to_dict.test.ts
Comment (dumbmatter) [inline __tests__/utils.set_to_dict.test.ts:1]: Update filename to match function

File: public/Face.tsx [hunk 14]
```diff
@@ -0,0 +1,86 @@
...
+      )
+      .join(",");
+    if (cache.has(serializedArgs)) {
+      return cache.get(serializedArgs);
+    }
+    const result = fn(...args);
+    cache.set(serializedArgs, result);
+    return result;
+  };
+};
```
Comment (dumbmatter) [inline public/Face.tsx:39]: It would probably be better to use https://react.dev/reference/react/memo rather than this custom memoization function.

You might be able to just wrap Face in React.memo and have it work. If the objects passed as props to Face (such as faceConfig) are the same (in terms of identity, not necessarily a deep compare) then  it won't re-render. So then if you create a new faceConfig object if (and only if) you edit some property inside it, then React.memo will work with no custom compare function and also be fast.

But also, why is memoization useful here? Are we re-rendering the same exact face many times? Why?

File: public/EditorPage.tsx [hunk 15]
```diff
@@ -0,0 +1,1105 @@
...
+      })}
+    </div>
+  );
+};
+
+const MainFaceDisplay = ({
+  modalDisclosure,
+}: {
+  modalDisclosure: any;
+}): JSX.Element => {
```
Comment (dumbmatter) [inline public/EditorPage.tsx:265]: Don't need to specify JSX.Element as a return type, TypeScript will infer it.

File: public/utils.ts [hunk 16]
```diff
@@ -0,0 +1,262 @@
...
+
+  Object.keys(flattenedObj)
+    .sort()
+    .forEach((key) => {
+      // @ts-ignore
+      returnString += `${key}: ${flattenedObj[key]}\n`;
+    });
+
+  return returnString;
+};
```
Comment (dumbmatter) [inline public/utils.ts:176]: Maybe a better way to do this - doesn't require flattening the object firs https://stackoverflow.com/a/53593328

```
function JSONstringifyOrder(obj, space)
{
    const allKeys = new Set();
    JSON.stringify(obj, (key, value) => (allKeys.add(key), value));
    return JSON.stringify(obj, Array.from(allKeys).sort(), space);
}
```

File: public/utils.ts [hunk 17]
```diff
@@ -0,0 +1,262 @@
...
+    returnArray.push(roundTwoDecimals(i));
+  }
+
+  return returnArray;
+};
+
+export const distinct = (arr: any[]): any[] => {
+  // @ts-ignore
+  return [...new Set<any>(arr)];
+};
```
Comment (dumbmatter) [inline public/utils.ts:139]: Better for TypeScript:

```
export const distinct = <T>(arr: T[]): T[] => {
  return [...new Set(arr)];
};
```

Then the return type is the same as the input type, rather than any[]

File: public/EditorPage.tsx
Comment (dumbmatter) [inline public/EditorPage.tsx:1]: Might make it a bit easier to read if this was split into multiple files

File: svgs/head/head10.svg
Comment (dumbmatter) [inline svgs/head/head10.svg:1]: Still many altered SVG files are here..

**Review by dumbmatter — CHANGES_REQUESTED** [review]
Just a few more comments :)

File: svgs/head/head10.svg
Comment (tomkennedy22) [reply svgs/head/head10.svg:1]: Fixed

File: public/EditorPage.tsx
Comment (tomkennedy22) [reply public/EditorPage.tsx:1]: I split some helper functions out

File: public/Face.tsx (hunk 14 above)
Comment (tomkennedy22) [reply public/Face.tsx:40]: Memoization helps a bunch upon rending second, third, etc options on the page. Each face is built by copying the current FaceConfig from Zustand store, and running Override on it with the various Gallery options. 

So Memoization is helpful when you select skin color A, select skin color B, then go back to skin color A - we wouldn't need to do any calculations at that point. You can feel this especially impactful with sliders - if you slide all the way to the right, and then all the way back to original point, you can feel the rendering be a tick quicker on the way back

My worry about custom vs default memoization, and basically everywhere I use `objStringifyInOrder`, is the nested values being caught properly, being able to string-compare objs, and not having to worry about if we did a deep or shallow copy prior

I'm happy to test this a bit more, but ultimately went with the annoying custom memoization after running into problems originally

File: public/utils.ts (hunk 16 above)
Comment (tomkennedy22) [reply public/utils.ts:175]: Somewhat addressed this in https://github.com/zengm-games/facesjs/pull/35#discussion_r1559782144

Comment (dumbmatter) [reply public/utils.ts:175]: I'm don't think that comment changes anything with respect to the specific implementation of stringifying JSON while accounting for the order of object keys.

File: public/Face.tsx (hunk 14 above)
Comment (dumbmatter) [reply public/Face.tsx:40]: In very unscientific testing I did, it doesn't seem faster to switch to a previously-selected option than to a new option. Are you seeing different behavior?

If we keep memoization, we should do something to avoid it staying in memory forever, in case someone is on the editor for a long time that'd be a memory leak.

But I think if we're able to lazy load the Face components, that might have a bigger impact than memoization, and might make memoization unnecessary. So I think that's the way to proceed... lazy load the rendering of Face components, and see if maybe that just solves all performance issues.

File: package.json [hunk 18]
```diff
@@ -34,37 +14,90 @@
...
-    "lint-staged": ">=10",
-    "prettier": "^3.1.0",
-    "rollup": "^4.5.0",
-    "svgo": "^3.0.2",
-    "typescript": "^5.2.2",
-    "ws": "^8.12.1"
+    "framer-motion": "^11.0.25",
+    "fs": "0.0.1-security",
+    "jest": "^29.7.0",
+    "path": "^0.12.7",
```
Comment (dumbmatter) [inline package.json:64]: We don't want this, we want the path built-in. Which you can reference as "node:path" to make it unambiguous when importing, and delete this erroneous dependency.

File: svgs/jersey/football5.svg
Comment (dumbmatter) [inline svgs/jersey/football5.svg:1]: This shouldn't be edited

File: public/EditorPage.tsx
Comment (dumbmatter) [inline public/EditorPage.tsx:1]: On the default gallery size, if I click one of the  "Randomize" buttons (for example next to "Choose Skin Color"), it might randomize it to a skin color that is currently not visible because of scroll position. In that case, it should scroll so the selected one is visible. https://developer.mozilla.org/en-US/docs/Web/API/Element/scrollIntoView might help with that.

Comment (dumbmatter) [inline public/EditorPage.tsx:1]: In the gallery for each feature, it should hide scrollbars that are useless (like in the default gallery size, the vertical scroll is never used, but it's shown as disabled always).

File: public/Face.tsx [hunk 19]
```diff
@@ -0,0 +1,62 @@
...
+  className?: string;
+  lazyLoad?: boolean;
+}) => {
+  const [ref, inView] = useInView({
+    triggerOnce: true,
+    threshold: 0,
+  });
+
+  const faceSvg =
+    inView || !lazyLoad ? faceToSvgString(faceConfig, overrides) : "";
```
Comment (dumbmatter) [inline public/Face.tsx:37]: It's probably better to use `display` rather than `faceToSvgString` - the latter is only intended for use by the CLI for exporting an SVG without a browser. Probably `display` is much faster. Although if not it would be interesting!

Also the way you have it now, it's generating the SVG every time the Face component renders. It'd be better to not do that. For `display` you'd generally put it in a `useEffect` that runs only the first time the component is rendered (or in this case, I guess the first time it's in view).

File: public/Face.tsx [hunk 20]
```diff
@@ -0,0 +1,62 @@
...
+  overrides?: Overrides;
+  maxWidth?: number;
+  width?: number;
+  className?: string;
+  lazyLoad?: boolean;
+}) => {
+  const [ref, inView] = useInView({
+    triggerOnce: true,
+    threshold: 0,
+  });
```
Comment (dumbmatter) [inline public/Face.tsx:34]: Is this doing what we want?

Imagine you scroll all the way everywhere, so all SVGs get loaded. Then you change a feature. That will then re-render all the SVGs, right? But we don't want that, we want only the visible ones to update. The ones offscreen should either remain static until they come in screen, or maybe even be deleted (replaced by a placeholder empty div) when off screen, if that matters for performance.

File: public/App.tsx [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+import { EditorPage } from "./EditorPage";
+
+const App = () => {
+  return (
+    <NextUIProvider>
+      <BrowserRouter>
+        <Routes>
+          <Route path="/" element={<Home />} />
+          <Route path="/editor" element={<EditorPage />} />
+          <Route path="/editor/:param" element={<EditorPage />} />
```
[the remark this answers is not in the document]
Comment (dumbmatter) [reply public/App.tsx:14]: Is the intention to support both the old and new formats? I'd rather just support the old one, having two is overkill. And the new one is more human readable... but nobody is reading a JSON string compressed into one line that is that long :)

**Review by dumbmatter — CHANGES_REQUESTED** [review]

File: public/shuffleFace.ts [hunk 22]
```diff
@@ -0,0 +1,53 @@
...
+      !gallerySectionConfig.randomizeEnabled &&
+      gallerySectionConfig.key === "body.color" &&
+      faceConfig.body.color &&
+      raceBySkinColor[faceConfig.body.color]
+    ) {
+      let inferredRace: Race = raceBySkinColor[
+        faceConfig.body.color
+      ][0] as Race;
+      if (inferredRace) {
+        options.race = inferredRace;
```
Comment (dumbmatter) [inline public/shuffleFace.ts:35]: Not sure we want this, it should be an explicit feature to select the race.

**Review by dumbmatter — CHANGES_REQUESTED** [review]

File: public/Face.tsx (hunk 14 above)
Comment (tomkennedy22) [reply public/Face.tsx]: Added lazy loading

File: public/App.tsx (hunk 21 above)
[the remark this answers is not in the document]
Comment (tomkennedy22) [reply public/App.tsx:14]: Flipped back to `btoa`

File: public/shuffleFace.ts (hunk 22 above)
Comment (tomkennedy22) [reply public/shuffleFace.ts:35]: Should this be a feature we add to the FaceConfig? I know on previous site it wasn't added into Face type and just tracked via dropdown. Curious on your take for how it should be tracked (maybe Race as well?)

File: public/EditorPage.tsx [hunk 23]
```diff
@@ -0,0 +1,838 @@
...
+  } else {
+    return <> </>;
+  }
+};
+
+// const scrollToFace = (ref: React.RefObject<HTMLDivElement>) => {
+//     if (ref.current) {
+//         ref.current!.scrollIntoView({ behavior: 'smooth', inline: 'center' });
+//     }
+// };
```
[the remark this answers is not in the document]
Comment (tomkennedy22) [reply public/EditorPage.tsx:523]: Keeping the refs as it helps w/ using display() + the lazyLoading, but otherwise dropping commented scroll code

File: public/Face.tsx (hunk 20 above)
Comment (tomkennedy22) [reply public/Face.tsx]: Switched to `triggerOnce: false` so it'll detach upon scroll out of view

File: svgs/jersey/football5.svg
Comment (tomkennedy22) [reply svgs/jersey/football5.svg:1]: I'm not actually sure what the problem is - spacing all looks the same?

File: public/EditorPage.tsx (hunk 23 above)
[the remark this answers is not in the document]
Comment (dumbmatter) [reply public/EditorPage.tsx:523]: You shouldn't need forwardRef (since there is no other need for accessing that ref outside of Face), so you can just define a ref inside Face.

File: public/shuffleFace.ts (hunk 22 above)
Comment (dumbmatter) [reply public/shuffleFace.ts:35]: Maybe, but for this PR I definitely don't want to change that.

File: svgs/jersey/football5.svg
Comment (dumbmatter) [reply svgs/jersey/football5.svg:1]: Newline at the end of the file :)

File: public/EditorPage.tsx [hunk 24]
```diff
@@ -0,0 +1,838 @@
...
+      shadow="md"
+      size="xl"
+      isOpen={isOpen}
+      onOpenChange={onOpenChange}
+    >
+      <ModalContent>
+        {(_) => (
+          <>
+            <ModalHeader className="flex flex-col gap-1">
+              Paste JSON to Render Face
```
[the remark this answers is not in the document]
Comment (tomkennedy22) [reply public/EditorPage.tsx:796]: Sounds good, changes made. Other thoughts:
- Should we validate that all features are included as user enters? Could run into issues where user enters empty JSON obj and gets empty face
- If they don't include a particular feature, should we not include that feature, or pick random value for it? Or maybe just get current feature value from currently showing face?
- Do you care about auto-formatting the JSON or anything?

[the remark this answers is not in the document]
Comment (dumbmatter) [reply public/EditorPage.tsx:796]: > Should we validate that all features are included as user enters? Could run into issues where user enters empty JSON obj and gets empty face

Might be nice, but I think it's fine to leave that for later.

> If they don't include a particular feature, should we not include that feature, or pick random value for it? Or maybe just get current feature value from currently showing face?

Probably would be best to keep the value from the current face. But either of the other options would be fine, whatever's easiest. Although I think "not include that feature" might just crash it.

> Do you care about auto-formatting the JSON or anything?

No

[the remark this answers is not in the document]
Comment (tomkennedy22) [reply public/EditorPage.tsx:796]: > Although I think "not include that feature" might just crash it.

As I just discovered in testing :D

File: svgs/jersey/football5.svg
Comment (tomkennedy22) [reply svgs/jersey/football5.svg:1]: 👍

**Review by dumbmatter — CHANGES_REQUESTED** [review]
When randomizing with only female selected, it has a decent chance of giving this specific beard, maybe due to the unneeded changes to core files I was worried about :)

image

File: public/EditorPage.tsx [hunk 25]
```diff
@@ -0,0 +1,895 @@
...
+                    >
+                      Customize Shuffle
+                    </p>
+                    <div className="flex gap-4">
+                      <CheckboxGroup
+                        label="Gender"
+                        isInvalid={genderInvalidOptions}
+                        value={shuffleGenderSettingObject}
+                        errorMessage={
+                          genderInvalidOptions ? "Select at least one" : null
```
Comment (dumbmatter) [inline public/EditorPage.tsx:761]: If none is selected, it still lets you shuffle, seemingly with the default value (male and all races) - maybe would be better to say "Will use default if none are selected" or something.

File: public/EditorPage.tsx [hunk 26]
```diff
@@ -0,0 +1,895 @@
...
+    (shuffleGenderSettingObject.length + shuffleRaceSettingObject.length);
+
+  return (
+    <div className="bg-slate-800 text-white flex justify-between w-full fixed z-50	">
+      <div className="flex gap-4 text-xl p-2 justify-around items-center">
+        <span className="cursor-pointer rounded-full p-1 m-0.5 hover:bg-slate-50 hover:text-slate-900">
+          <House weight="fill" size={24} onClick={() => navigate("/")} />
+        </span>
+        <span className="hidden md:inline">faces.js Editor</span>
+        <Badge
```
Comment (dumbmatter) [inline public/EditorPage.tsx:722]: At a glance I'm not sure exactly where the code is for it... but that little circle with a number in it next to the settings... I don't know what that's supposed to represent, so maybe it should just be deleted.

**Review by dumbmatter — APPROVED** [review]

**Comment by tomkennedy22** [comment]
> I think I'm done working on this... could keep going forever, but I think this is pretty good. @tomkennedy22 please take a look at it and let me know what you think - definitely possible I broke something! And let me know if there's anything else you'd like to do here, or if you think it's basically done too.
> 
> Then next step would be to ask a few people to test it and see what they say.

I think it looks great!

**Comment by dumbmatter** [comment]
Actually the color gallery faces were broken! Fixed now, I'll put a beta version online soon
